### DISCUSSION TITLE
Why is <NavigationMenu.List> wrapped in a div with relative positioning?

### DISCUSSION THREAD

**Discussion by adatti** [radix-ui/primitives#1874] [Help]
I just noticed that `NavigationMenu.List` isn't rendered simply as a `ul` but as a `div style="position:relative"` and a `ul` inside it. The div seems out of reach, because the attributes I set on the `NavigationMenu.List` only apply to the `ul`. This can be seen by inspecting the element in [this sandbox](https://codesandbox.io/s/quirky-fermi-rm4ptl?file=/src/App.tsx) or on the [radix navigation menu demo page](https://www.radix-ui.com/docs/primitives/components/navigation-menu).

![image](https://user-images.githubusercontent.com/98817622/210991958-4a98f244-0f8e-4e32-a4cb-7be514eaabda.png)

Is there a reason for this? I only just noticed this and it kind of breaks my ~heart~ layout. I can get the div to be ignored by my CSS by adding `display: contents` to it, but I have to select it awkwardly with `.nav > *:nth-child(2)` (because the div happens to be the second child of the nav).

**Comment by andy-hook** [comment]
@adatti This is used as a positioning "track" for `NavigationMenu.Arrow`.

https://user-images.githubusercontent.com/11708259/211013609-b2bc17ea-d779-4aaf-b7dc-4e2c0deb7cba.mp4

We decided that wrapping was acceptable here as you'd typically apply your list positioning to `Root` as this directly wraps the list. The only time I think this wouldn't be the case is when you're trying to nest other elements next to `List` within `Root`, is that the case for you? do you have an example of your usage?

**Reply by deltasierra96** [reply]
> @adatti This is used as a positioning "track" for `NavigationMenu.Arrow`.
> 
>  CleanShot.2023-01-06.at.12.34.31.mp4 
> We decided that wrapping was acceptable here as you'd typically apply your list positioning to `Root` as this directly wraps the list. The only time I think this wouldn't be the case is when you're trying to nest other elements next to `List` within `Root`, is that the case for you? do you have an example of your usage?

I have an example that I believe what @adatti is having.

With the orientation of the sub menu set to vertical, the `NavigationMenu.Trigger` has its height calculated and applied to the  `NavigationMenu.Indicator` just fine. However if I wanted to set the width of it to be full width of the tab control grid column, it overflows.

![preview](https://user-images.githubusercontent.com/33609328/211174004-68356447-7481-4432-a51d-e2bd89ad9290.png)

```jsx
<NavigationMenu.Sub orientation="vertical" defaultValue="sub1">
  {/* NavigationMenu.List */}
  <NavigationMenu.List>
    <div className="grid grid-flow-col grid-cols-12 gap-2">
      {/* Tab column */}
      <div className="col-span-3 relative">
        {/* Item */}
        <NavigationMenu.Item value="sub1">
          <NavigationMenu.Trigger className="w-full p-4 rounded-md ">
            <span className="block font-medium">Sub Item One</span>
            <p className="text-sm">
              Accept payments online, in person, or through your platform.
            </p>
          </NavigationMenu.Trigger>
          <NavigationMenu.Content
            className={clsx(
              "absolute inset-x-0 w-full",
              "radix-content",
              "radix-motion-from-start:animate-enter-from-top",
              "radix-motion-from-end:animate-enter-from-bottom",
              "radix-motion-to-start:animate-exit-to-top",
              "radix-motion-to-end:animate-exit-to-bottom"
            )}
          >
            <div className="w-full p-3 ">
              <div className="grid grid-cols-6 gap-4">
                <div className="w-full col-span-2 p-4 bg-gray-100 rounded-md dark:bg-gray-900"></div>

                <div className="flex flex-col w-full col-span-4 p-4 space-y-3 bg-gray-100 rounded-md dark:bg-gray-900">
                  <div className="w-full h-12 bg-white rounded-md dark:bg-gray-700"></div>
                  <div className="w-full h-12 bg-white rounded-md dark:bg-gray-700"></div>
                  <div className="w-full h-12 bg-white rounded-md dark:bg-gray-700"></div>
                  <div className="w-full h-12 bg-white rounded-md dark:bg-gray-700"></div>
                </div>
              </div>
            </div>
          </NavigationMenu.Content>
        </NavigationMenu.Item>
        {/* Item */}
        <NavigationMenu.Item value="sub2">
          <NavigationMenu.Trigger className="w-full p-4 rounded-md ">
            <span className="block font-medium">Sub Item Two</span>
            <p className="text-sm">
              Accept payments online, in person, or through your platform.
            </p>
          </NavigationMenu.Trigger>
          <NavigationMenu.Content
            className={clsx(
              "absolute inset-x-0 w-full",
              "radix-content",
              "radix-motion-from-start:animate-enter-from-top",
              "radix-motion-from-end:animate-enter-from-bottom",
              "radix-motion-to-start:animate-exit-to-top",
              "radix-motion-to-end:animate-exit-to-bottom"
            )}
          >
            <div className="w-full p-3 ">
              <div className="grid grid-cols-6 gap-4">
                <div className="w-full col-span-2 p-4 bg-gray-100 rounded-md dark:bg-gray-900"></div>

                <div className="flex flex-col w-full col-span-4 p-4 space-y-3 bg-gray-100 rounded-md dark:bg-gray-900">
                  <div className="w-full h-12 bg-white rounded-md dark:bg-gray-700"></div>
                  <div className="w-full h-12 bg-white rounded-md dark:bg-gray-700"></div>
                  <div className="w-full h-12 bg-white rounded-md dark:bg-gray-700"></div>
                  <div className="w-full h-12 bg-white rounded-md dark:bg-gray-700"></div>
                </div>
              </div>
            </div>
          </NavigationMenu.Content>
        </NavigationMenu.Item>
        {/* Indicator */}
        <NavigationMenu.Indicator className="w-full bg-white rounded-md will-change-[transform,height]" />
      </div>
      {/* Viewport column */}
      <NavigationMenu.Viewport className="relative col-span-9 overflow-hidden ease-easy-it-goes duration-200 will-change-[transform,_width,_height] bg-white h-radix-navigation-menu-viewport" />
    </div>
  </NavigationMenu.List>
</NavigationMenu.Sub>;

```

Could we ultimately not allow the `NavigationMenu.Indicator` to be played anywhere within the `NavigationMenu.List` component and handle our own relative positioning?

**Comment by heymath** [comment]
This added `div` with a `position: relative;` is messing with our nav styles.

We don't have the exact same positionning / logic as the default navigation style so we have to add a rule that looks like it :

```css
.nav > div {
     position: initial !important;
}
```

And it can mess with your style when your nav is in display flex. Even more when you add 2 List as children for instance but then you have a div that actually wraps these children so the styles you add on your nav-list should maybe be on the div position:relative but you have no way to add it in a clean way, sadly.

```tsx
<Root className="nav">
    <List className="nav-list">...</List>
    <List className="nav-list">...</List>
</Root>
```

It would be nice to not have to style our nav (actually reset Radix styles) using these kind of dangerous / too vague selectors.

Why do not add ListRelativeWrapper or I don't know what name/logic it would have so users can manually add it or not?
Because from issues opened about NavigationMenu, it looks like some people do not want this default behavior.

**Reply by minhwpm** [reply]
I agree, the <NavigationMenu.List> should not have that relative positioning.
It mess up the styling flexibility which is the most valuable of Radix UI.

**Comment by Balros** [comment]
Another workaround is to use `display: grid; width: 100%; height: 100%;` in parent `Root` element.
```JavaScript
<Root style="display:grid; width: 100%; height: 100%;">
    <List className="nav-list">...</List>
    <List className="nav-list">...</List>
</Root>
```

**Comment by sleeyax** [comment]
Still an issue today. There's no way to opt out of adding this weird div in my markup without applying the suggested CSS hack. Radix is usually very composable, why suddenly deviate from that goal here?

**Reply by SalahAdDin** [reply]
> Still an issue today. There's no way to opt out of adding this weird div in my markup without applying the suggested CSS hack. Radix is usually very composable, why suddenly deviate from that goal here?

Which CSS hack?

**Reply by sleeyax** [reply]
> > Still an issue today. There's no way to opt out of adding this weird div in my markup without applying the suggested CSS hack. Radix is usually very composable, why suddenly deviate from that goal here?
> 
> Which CSS hack?

`display: contents;` as suggested in the top post.

Sure, maybe calling it a hack is a bit of a reach but it's certainly not a clean solution. The div just shouldn't be there when I don't need it to.

**Comment by berkerdemirer** [comment]
So, if i want to have justify between inside the list, I can't because i can't set the list width to be 100% as I can't target that.

**Reply by SalahAdDin** [reply]
> So, if i want to have justify between inside the list, I can't because i can't set the list width to be 100% as I can't target that.

It seems to be like that.

**Reply by berkerdemirer** [reply]
Quite limiting tbh

**Comment by alex-r89** [comment]
Just spent 2 hours trying to work out where this div was coming from and how to remove/edit it. Shame that I need to add hacky CSS to remove the styling

**Reply by SalahAdDin** [reply]
could you share yours olution?

**Reply by AanshKot** [reply]
I had the same problem, fixed it using the styling hack they mentioned, added it to my globals.css. `nav > div:first-child {
	display: contents
}`

**Comment by codePassion-dot** [comment]
I still don't understand why a "headless" library is adding styles because that would potentially be the avg behavior. From what I understand headless library should only provide functionality and accessibility nothing related to styles.

**Reply by codePassion-dot** [reply]
headless library should not think about typical behavior for styles, who cares ? let the consumer style it however they need/want

**Comment by PAXANDDOS** [comment]
This issue is 2 years old now?! I'm glad I found a workaround here because I was about to lose my mind. Imagine spending 2 hours trying to apply "ml-auto" to a single element of the list

**Reply by legnaib** [reply]
Similar for me, just tried to add "justify-between" which was not working until finding this odd div. Annoying that this is unnecessarily added

**Comment by pallaoro** [comment]
Still there

**Comment by mohitkumawat310** [comment]
I just replaced it with a standard < ul> tag. It got fixed with this.

**Reply by SalahAdDin** [reply]
any snippet?

**Reply by mohitkumawat310** [reply]
I am using shadcn UI. I removed the wrapper (NavigationMenuList) and replaced it with a normal <ul> tag.

`<ul className="flex justify-start">....</ul>`

This fixed the alignment issue with the navigation content in the mega menu for me.

**Comment by tostesdaniel** [comment]
For anyone using shadcn/ui Navigation Menu, this does the trick.

_> globals.css_
```tsx
@layer utilities {
  nav[data-slot="navigation-menu"] > div:first-child {
    display: contents;
  }
}
```

**Inspired by this reply:** https://github.com/radix-ui/primitives/discussions/1874#discussioncomment-12595545
